File: aoc2024/aoc25.hs
import Data.Either (lefts, rights)
import Data.List.Split (splitOn)
import System.Environment (getArgs)

newtype Lock = Lock [Int] deriving (Show)

newtype Key = Key [Int] deriving (Show)

countPounds :: [String] -> [Int]
countPounds =
  foldr
    (\v -> zipWith (+) [if ch == '#' then 1 else 0 | ch <- v])
    (repeat 0)

parsePara :: [String] -> Either Lock Key
parsePara [] = error "empty para"
parsePara para | all (== '#') (head para) = Left $ Lock $ countPounds (tail para)
parsePara para | all (== '#') (last para) = Right $ Key $ countPounds (init para)
parsePara para = error $ "Bad para: " ++ show para

fits :: Lock -> Key -> Bool
fits (Lock ls) (Key ks) = maximum (zipWith (+) ls ks) < 6

main :: IO ()
main = do
  args <- getArgs
  let filename = if null args then "aoc25.in" else head args
  pieces <- splitOn "\n\n" <$> readFile filename
  let parsed = parsePara . lines <$> pieces
  putStr "Part 1: "
  print $ length [() | l <- lefts parsed, k <- rights parsed, fits l k]

File: aoc2024/aoc25.py
import re
from aoc_util import get_data_paras

data = get_data_paras(25)

locks = []
keys = []
for para in data:
    lines = para.splitlines()
    if re.match(r"^#+$", lines[0]):
        heights = [0] * len(lines[0])
        for line in lines[1:]:
            for hidx, ch in enumerate(line):
                if ch == "#":
                    heights[hidx] += 1
        locks.append(tuple(heights))
    elif re.match(r"^#+$", lines[-1]):
        # key
        heights = [0] * len(lines[0])
        for line in lines[:-1]:
            for hidx, ch in enumerate(line):
                if ch == "#":
                    heights[hidx] += 1
        keys.append(tuple(heights))
    else:
        print(f"ERROR! Bad para:\n\n{para}\n")
        exit(1)

total = 0
for lock in locks:
    for key in keys:
        if all(x + y < 6 for (x, y) in zip(lock, key)):
            total += 1
print("Part 1:", total)
